ci_watson/base_classes.py

```diff
@@ -0,0 +1,401 @@
+"""Regression test helpers."""
+import getpass
+import os
+import sys
+from io import StringIO
+import shutil
+import datetime
+from difflib import unified_diff
+
+import pytest
+from astropy.io import fits
+from astropy.io.fits import FITSDiff, HDUDiff
+from astropy.utils.data import conf
+
+from .artifactory_helpers import get_bigdata, generate_upload_schema
+
+
+# Base classes for actual tests.
+# NOTE: Named in a way so pytest will not pick them up here.
+class BaseTest(object):
```

Comment:
I wonder if this should be here at all or better off as part of spacetelescope/jwst#2529 . While I appreciate your work to make it more general and put it in this PR, it appears to be too specialized to be used outside of `jwst` pipeline tests. For instance, it makes assumption on how exactly the big data will be laid out.

For instance, compare this with the one used in HSTCAL (starting at https://github.com/spacetelescope/hstcal/blob/ed4e9e90845ae9d281fec0cb26aba41a5433dae1/tests/helpers.py#L126). HST does not use half of these options.

We have to ask, is this ever going to be useful in a package outside of `jwst`? If you must put it in `ci-watson`, I propose this to be moved to `jwst_helpers.py` instead. Or we can even defer this portion as a follow-up PR in the future.

Some of the methods here, though, like generating upload schema can be a standalone function perhaps, which then a specialized test class can call in its own method.

Reply:
I agree that it would be better to have these in the `jwst` repo for now.  It would also make the changes in spacetelescope/jwst#2529 possible to test using Jenkins as well.  Currently Jenkins can only use `ci_watkins` releases on pypi.  In the future we can always push functionality up from `jwst` to `ci_watson`, once it has proved itself in the `jwst` environment.

I also like the idea of any general-use helpers in `ci_watson` being standalone functions or `pytest` fixtures that can be imported *a la cart* to any tests that need them.

Reply:
This base class was actually derived from the `HSTCAL` class and made MORE general, which accounts for the additional options.  In short, the implementation in `ci_watson` provides FULL control over all available options for `FITSDiff` and `HDUDiff`, unlike the limited implementation used in `HSTCAL`.  The `HSTCAL` classes could, in priniciple, actually be based on the `ci_watson` base class with minimal or no changes.  In fact, I have plans to migrate all the `Drizzlepac` tests to use the `ci_watson` base class given here as is, work which will take place in Oct as part of the HLA in the Pipeline Project.  Does this serve to demonstrate its usefulness outside of `jwst`?  Providing this base class was intended to limit the number of implementations of common code useful for all missions regression tests (or at least, something that can be revised and maintained to support multiple missions).  

The initial implementation of `compare_outputs()` in `ci_watson.base_classes.BaseTest` also serves as the default implementation of the mechanism (already used by DrizzlePac) for OK'ing new test results.  It demonstrates how to set up and use the currently available function `generate_upload_schema()`.  However, the setup information for this function still needs to be turned into its own function which would be necessary in order to use the generate function properly. It should be possible to create a stand-alone function (or functions) to reproduce this behavior, but I would suggest that work wait until we actually have a use-case for the functional form as opposed to what is already working for all the JWST tests.

Reply:
We discussed that `BaseTest` should live in the `jwst` repository.  But it can call general functions such as `compare_outputs` that can live in `ci_watson`.

And in fact, a `BaseJWSTTest` already exists in the `jwst` repository in PR spacetelescope/jwst#2529, so you may want to just move the attributes from `BaseTest` down to `BaseJWSTTest` that and get rid of `BaseTest` altogether.  I would also put something into the class name that it is specifically for regression testing and fitsdiffing.  It is not, for example, the base of our unit tests, or of even all our regression tests.